crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -20,41 +22,64 @@ pub struct DisplayType<'db> {
     db: &'db dyn Db,
 }
 
-impl Display for DisplayType<'_> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+impl DisplayType<'_> {
+    /// Returns the string representation of a type, which is the value displayed either as
+    /// `Literal[<repr>]` or `Literal[<repr1>, <repr2>]` for literal types or as `<repr>` for non
+    /// literals
+    fn representation(&self) -> String {
```

Comment:
Returning a `String` here results in an unnecessary allocation. What we can do to avoid the string allocation is to change the method to `write_representation(&self, f: &mut Formatter) -> std::fmt::Result` and directly write into the target output string.